File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        a,b = len(s),len(t)
        # Recursion
        # def find(i,j):
        #     if j<0:
        #         return 1
        #     if i<0:
        #         return 0

        #     if s[i]==t[j]:
        #         return find(i-1,j-1)+find(i-1,j)
        #     else:
        #         return find(i-1,j)
        # return find(a-1,b-1)

        #Memoization
        # dp = [[-1]*(b) for _ in xrange(a)]
        # def find(i,j):
        #     if j<0:
        #         return 1
        #     if i<0:
        #         return 0
        #     if dp[i][j]!=-1:
        #         return dp[i][j]

        #     if s[i]==t[j]:
        #         dp[i][j] = find(i-1,j-1)+find(i-1,j)
        #         return dp[i][j]
            
        #     dp[i][j] = find(i-1,j)
        #     return dp[i][j]
        # return find(a-1,b-1)

        #Tabulation
        # dp = [[0]*(b+1) for _ in xrange(a+1)]
        # #Base case 
        # for i in xrange(a+1):
        #     dp[i][0] =  1
        
        # for i in xrange(1,a+1):
        #     for j in xrange(1,b+1):
        #         if s[i-1] == t[j-1]:
        #             dp[i][j] = dp[i-1][j-1] + dp[i-1][j]
        #         else:
        #             dp[i][j] = dp[i-1][j]
        
        # return dp[a][b]

        #Space Optimization
        prev = [0]*(b+1)
        prev[0] = 1

        for i in xrange(1,a+1):
            cur = [0]*(b+1)
            cur[0] = 1
            for j in xrange(1,b+1):
                if s[i-1] == t[j-1]:
                    cur[j] = prev[j-1] + prev[j]
                else:
                    cur[j] = prev[j]
            prev = cur

        return prev[b]

File: 0115-distinct-subsequences/0115-distinct-subsequences.java
import java.util.Arrays;

class Solution {
    public int numDistinct(String s, String t) {
        int m = s.length(), n = t.length();
        
        // If t is longer than s, no valid subsequence exists
        if (n > m) {
            return 0;
        }
        
        int [] prev = new int [n + 1];
        prev[0] = 1;
        
        for (int i = 1; i < m + 1; i++){
            int [] cur = new int [n + 1];
            cur[0] = 1;
            for (int j = 1; j < n + 1; j++){
                if (s.charAt(i - 1) == t.charAt(j - 1)){
                    cur[j] = prev[j - 1] + prev[j];
                }
                else{
                    cur[j] = prev[j];
                }
            }
            prev = cur;
        }
        return prev[n];
        
//         int[][] dp = new int[m][n];
//         for (int[] rows : dp) {
//             Arrays.fill(rows, -1);
//         }
        
//         return helper(m - 1, n - 1, s, t, dp);
    }

    private int helper(int i, int j, String s, String t, int[][] dp) {
        if (j < 0) {
            // If t is exhausted, there is exactly one way to match (by taking no characters from s)
            return 1;
        }
        if (i < 0) {
            // If s is exhausted but t is not, there is no way to match
            return 0;
        }
        if (dp[i][j] != -1) {
            // Return already computed result
            return dp[i][j];
        }
        if (s.charAt(i) == t.charAt(j)) {
            // If characters match, choose to include or exclude the character
            dp[i][j] = helper(i - 1, j - 1, s, t, dp) + helper(i - 1, j, s, t, dp);
        } else {
            // If characters don't match, exclude the character from s
            dp[i][j] = helper(i - 1, j, s, t, dp);
        }
        return dp[i][j];
    }
}
